File: electrical/src/main/scala/resonantinduction/electrical/multimeter/PartMultimeter.java
package resonantinduction.electrical.multimeter;

import java.util.HashSet;
import java.util.Set;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.MovingObjectPosition;
import net.minecraftforge.common.ForgeDirection;
import net.minecraftforge.fluids.FluidTankInfo;
import net.minecraftforge.fluids.IFluidHandler;
import resonant.api.grid.INodeProvider;
import resonant.lib.network.IPacketReceiver;
import resonant.lib.utility.WrenchUtility;
import resonantinduction.core.ResonantInduction;
import resonantinduction.core.grid.fluid.FluidPressureNode;
import resonantinduction.core.grid.fluid.IPressureNodeProvider;
import resonantinduction.core.interfaces.IMechanicalNode;
import resonantinduction.core.prefab.part.PartFace;
import resonantinduction.electrical.Electrical;
import universalelectricity.api.CompatibilityModule;
import universalelectricity.api.electricity.IElectricalNetwork;
import universalelectricity.api.energy.IConductor;
import universalelectricity.api.energy.IEnergyNetwork;
import universalelectricity.api.net.IConnector;
import codechicken.lib.data.MCDataInput;
import codechicken.lib.data.MCDataOutput;
import codechicken.lib.vec.Cuboid6;
import codechicken.lib.vec.Vector3;
import codechicken.multipart.IRedstonePart;
import codechicken.multipart.TMultiPart;
import codechicken.multipart.TileMultipart;

import com.google.common.io.ByteArrayDataInput;

import cpw.mods.fml.common.network.PacketDispatcher;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/** Block that detects power.
 * 
 * @author Calclavia */
public class PartMultimeter extends PartFace implements IConnector<MultimeterNetwork>, IRedstonePart, IPacketReceiver
{
    public enum DetectMode
    {
        NONE("none"),
        LESS_THAN("lessThan"),
        LESS_THAN_EQUAL("lessThanOrEqual"),
        EQUAL("equal"),
        GREATER_THAN_EQUAL("greaterThanOrEqual"),
        GREATER_THAN("greaterThan");

        public String display;

        private DetectMode(String s)
        {
            display = s;
        }
    }

    public Set<EntityPlayer> playersUsing = new HashSet<EntityPlayer>();

    /** Detection */
    public double redstoneTriggerLimit;
    public byte detectType = 0;
    public byte graphType = 0;
    private DetectMode detectMode = DetectMode.NONE;
    public boolean redstoneOn;
    private boolean doDetect = true;

    public boolean isPrimary;
    private MultimeterNetwork network;

    public boolean hasMultimeter(int x, int y, int z)
    {
        return getMultimeter(x, y, z) != null;
    }

    @Override
    public void preRemove()
    {
        if (!world().isRemote)
            getNetwork().split(this);
    }

    public void refresh()
    {
        if (world() != null)
        {
            if (!world().isRemote)
            {
                for (Object obj : getConnections())
                {
                    if (obj instanceof PartMultimeter)
                    {
                        getNetwork().merge(((PartMultimeter) obj).getNetwork());
                    }
                }

                getNetwork().reconstruct();
            }
        }
    }

    public void updateDesc()
    {
        writeDesc(getWriteStream());
    }

    public void updateGraph()
    {
        writeGraph(getWriteStream());
    }

    @Override
    public void onWorldJoin()
    {
        refresh();
    }

    @Override
    public void onNeighborChanged()
    {
        refresh();
    }

    @Override
    public void onPartChanged(TMultiPart part)
    {
        refresh();
    }

    /** Gets the multimeter on the same plane. */
    public PartMultimeter getMultimeter(int x, int y, int z)
    {
        TileEntity tileEntity = world().getBlockTileEntity(x, y, z);

        if (tileEntity instanceof TileMultipart)
        {
            TMultiPart part = ((TileMultipart) tileEntity).partMap(placementSide.ordinal());

            if (part instanceof PartMultimeter)
            {
                return (PartMultimeter) part;
            }
        }

        return null;
    }

    @Override
    public boolean activate(EntityPlayer player, MovingObjectPosition part, ItemStack item)
    {
        if (WrenchUtility.isUsableWrench(player, player.inventory.getCurrentItem(), x(), y(), z()))
        {
            if (!this.world().isRemote)
            {
                doDetect = !doDetect;
                player.addChatMessage("Multimeter detection set to: " + doDetect);
                WrenchUtility.damageWrench(player, player.inventory.getCurrentItem(), x(), y(), z());
            }
            return true;
        }

        player.openGui(Electrical.INSTANCE, placementSide.ordinal(), world(), x(), y(), z());
        return true;
    }

    @Override
    public void update()
    {
        super.update();

        this.ticks++;

        if (!world().isRemote)
        {
            if (doDetect)
                updateDetections();

            double detectedValue = getNetwork().graphs.get(detectType).getDouble();

            boolean outputRedstone = false;

            switch (detectMode)
            {
                default:
                    break;
                case EQUAL:
                    outputRedstone = detectedValue == redstoneTriggerLimit;
                    break;
                case GREATER_THAN:
                    outputRedstone = detectedValue > redstoneTriggerLimit;
                    break;
                case GREATER_THAN_EQUAL:
                    outputRedstone = detectedValue >= redstoneTriggerLimit;
                    break;
                case LESS_THAN:
                    outputRedstone = detectedValue < redstoneTriggerLimit;
                    break;
                case LESS_THAN_EQUAL:
                    outputRedstone = detectedValue <= redstoneTriggerLimit;
                    break;
            }

            getNetwork().markUpdate();

            if (ticks % 20 == 0)
            {
                if (outputRedstone != redstoneOn)
                {
                    redstoneOn = outputRedstone;
                    tile().notifyPartChange(this);
                }

                updateGraph();
            }
        }

        if (!world().isRemote)
        {
            for (EntityPlayer player : playersUsing)
            {
                updateGraph();
            }
        }
    }

    public void updateDetections()
    {
        ForgeDirection receivingSide = getDirection().getOpposite();
        TileEntity tileEntity = getDetectedTile();

        /** Update Energy Graph */
        if (tileEntity instanceof IConductor)
        {
            IConnector<IEnergyNetwork> instance = ((IConductor) tileEntity).getInstance(receivingSide);

            for (ForgeDirection dir : ForgeDirection.values())
            {
                if (instance != null)
                {
                    break;
                }

                instance = ((IConnector) tileEntity).getInstance(dir);
            }

            if (instance != null)
            {
                if (instance.getNetwork() instanceof IEnergyNetwork)
                {
                    IEnergyNetwork network = instance.getNetwork();
                    getNetwork().energyGraph.queue(Math.max(network.getBuffer(), network.getLastBuffer()));
                    getNetwork().powerGraph.queue(getNetwork().energyGraph.getAverage() * 20);

                    if (instance.getNetwork() instanceof IElectricalNetwork)
                        getNetwork().voltageGraph.queue(((IElectricalNetwork) network).getVoltage());
                }
            }
        }

        if (tileEntity instanceof INodeProvider)
        {
            IMechanicalNode instance = (IMechanicalNode) ((INodeProvider) tileEntity).getNode(IMechanicalNode.class, receivingSide);

            for (ForgeDirection dir : ForgeDirection.values())
            {
                if (instance != null)
                {
                    break;
                }

                instance = (IMechanicalNode) ((INodeProvider) tileEntity).getNode(IMechanicalNode.class, dir);
            }

            if (instance != null)
            {
                getNetwork().torqueGraph.queue(instance.getTorque());
                getNetwork().angularVelocityGraph.queue(instance.getAngularSpeed());
                getNetwork().powerGraph.queue((long) instance.getPower());
            }
        }

        if (tileEntity instanceof IFluidHandler)
        {
            FluidTankInfo[] fluidInfo = ((IFluidHandler) tileEntity).getTankInfo(receivingSide);

            if (fluidInfo != null)
            {
                for (FluidTankInfo info : fluidInfo)
                {
                    if (info != null)
                        if (info.fluid != null)
                            getNetwork().fluidGraph.queue(info.fluid.amount);
                }
            }
        }

        if (tileEntity instanceof IPressureNodeProvider)
        {
            getNetwork().pressureGraph.queue(((FluidPressureNode) ((IPressureNodeProvider) tileEntity).getNode(FluidPressureNode.class, receivingSide)).getPressure(receivingSide));
        }

        getNetwork().energyGraph.queue(CompatibilityModule.getEnergy(tileEntity, receivingSide));

        /** Update Energy Capacity Graph */
        getNetwork().energyCapacityGraph.queue(CompatibilityModule.getMaxEnergy(tileEntity, receivingSide));
    }

    @Override
    public void readDesc(MCDataInput packet)
    {
        packet.readByte();
        placementSide = ForgeDirection.getOrientation(packet.readByte());
        facing = packet.readByte();
        detectMode = DetectMode.values()[packet.readByte()];
        detectType = packet.readByte();
        graphType = packet.readByte();
        getNetwork().center = new universalelectricity.api.vector.Vector3(packet.readNBTTagCompound());
        getNetwork().size = new universalelectricity.api.vector.Vector3(packet.readNBTTagCompound());
        getNetwork().isEnabled = packet.readBoolean();
    }

    @Override
    public void writeDesc(MCDataOutput packet)
    {
        packet.writeByte(0);
        packet.writeByte(placementSide.ordinal());
        packet.writeByte(facing);
        packet.writeByte(detectMode.ordinal());
        packet.writeByte(detectType);
        packet.writeByte(graphType);
        packet.writeNBTTagCompound(getNetwork().center.writeToNBT(new NBTTagCompound()));
        packet.writeNBTTagCompound(getNetwork().size.writeToNBT(new NBTTagCompound()));
        packet.writeBoolean(getNetwork().isEnabled);
    }

    public void writeGraph(MCDataOutput packet)
    {
        packet.writeByte(2);
        isPrimary = getNetwork().isPrimary(this);
        packet.writeBoolean(isPrimary);

        if (isPrimary)
            packet.writeNBTTagCompound(getNetwork().save());
    }

    @Override
    public void read(MCDataInput packet)
    {
        read(packet, packet.readUByte());
    }

    public void read(MCDataInput packet, int packetID)
    {
        switch (packetID)
        {
            case 0:
            {
                placementSide = ForgeDirection.getOrientation(packet.readByte());
                facing = packet.readByte();
                detectMode = DetectMode.values()[packet.readByte()];
                detectType = packet.readByte();
                graphType = packet.readByte();
                getNetwork().center = new universalelectricity.api.vector.Vector3(packet.readNBTTagCompound());
                getNetwork().size = new universalelectricity.api.vector.Vector3(packet.readNBTTagCompound());
                getNetwork().isEnabled = packet.readBoolean();
                refresh();
                break;
            }
            case 1:
            {
                redstoneTriggerLimit = packet.readLong();
                break;
            }
            case 2:
            {
                isPrimary = packet.readBoolean();

                if (isPrimary)
                    getNetwork().load(packet.readNBTTagCompound());
                break;
            }
        }
    }

    @Override
    public void onReceivePacket(ByteArrayDataInput data, EntityPlayer player, Object... extra)
    {
        detectMode = DetectMode.values()[data.readByte()];
        detectType = data.readByte();
        graphType = data.readByte();
        redstoneTriggerLimit = data.readDouble();
    }

    public TileEntity getDetectedTile()
    {
        ForgeDirection direction = getDirection();
        return world().getBlockTileEntity(x() + direction.offsetX, y() + direction.offsetY, z() + direction.offsetZ);
    }

    public ForgeDirection getDirection()
    {
        return ForgeDirection.getOrientation(this.placementSide.ordinal());
    }

    public void toggleGraphType()
    {
        graphType = (byte) ((graphType + 1) % getNetwork().graphs.size());
        updateServer();
    }

    public void toggleMode()
    {
        detectMode = DetectMode.values()[(detectMode.ordinal() + 1) % DetectMode.values().length];
        updateServer();
    }

    public void toggleDetectionValue()
    {
        detectType = (byte) ((detectType + 1) % getNetwork().graphs.size());
        updateServer();
    }

    public void updateServer()
    {
        PacketDispatcher.sendPacketToServer(ResonantInduction.PACKET_MULTIPART.getPacket(new universalelectricity.api.vector.Vector3(x(), y(), z()), placementSide.ordinal(), (byte) detectMode.ordinal(), detectType, graphType, redstoneTriggerLimit));
    }

    @Override
    public void load(NBTTagCompound nbt)
    {
        super.load(nbt);
        placementSide = ForgeDirection.getOrientation(nbt.getByte("side"));
        detectMode = DetectMode.values()[nbt.getByte("detectMode")];
        detectType = nbt.getByte("detectionType");
        graphType = nbt.getByte("graphType");
        doDetect = nbt.getBoolean("doDetect");
        redstoneTriggerLimit = nbt.getDouble("triggerLimit");
    }

    @Override
    public void save(NBTTagCompound nbt)
    {
        super.save(nbt);
        nbt.setByte("side", (byte) placementSide.ordinal());
        nbt.setByte("detectMode", (byte) detectMode.ordinal());
        nbt.setByte("detectionType", detectType);
        nbt.setByte("graphType", graphType);
        nbt.setBoolean("doDetect", doDetect);
        nbt.setDouble("triggerLimit", redstoneTriggerLimit);
    }

    public DetectMode getMode()
    {
        return detectMode;
    }

    @Override
    public String getType()
    {
        return "resonant_induction_multimeter";
    }

    @Override
    public int redstoneConductionMap()
    {
        return 0x1F;
    }

    @Override
    public boolean solid(int arg0)
    {
        return true;
    }

    @Override
    protected ItemStack getItem()
    {
        return new ItemStack(Electrical.itemMultimeter);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void renderDynamic(Vector3 pos, float frame, int pass)
    {
        if (pass == 0)
        {
            RenderMultimeter.render(this, pos.x, pos.y, pos.z);
        }
    }

    @Override
    public boolean canConnectRedstone(int arg0)
    {
        return true;
    }

    @Override
    public int strongPowerLevel(int arg0)
    {
        return redstoneOn ? 14 : 0;
    }

    @Override
    public int weakPowerLevel(int arg0)
    {
        return redstoneOn ? 14 : 0;
    }

    @Override
    public MultimeterNetwork getNetwork()
    {
        if (network == null)
        {
            network = new MultimeterNetwork();
            network.addConnector(this);
        }

        return network;
    }

    @Override
    public void setNetwork(MultimeterNetwork network)
    {
        this.network = network;
    }

    @Override
    public boolean canConnect(ForgeDirection direction, Object obj)
    {
        return obj instanceof PartMultimeter;
    }

    @Override
    public Object[] getConnections()
    {
        Object[] connections = new Object[6];

        for (ForgeDirection dir : ForgeDirection.VALID_DIRECTIONS)
        {
            if (dir != getDirection() && dir != getDirection().getOpposite())
            {
                universalelectricity.api.vector.Vector3 vector = getPosition().translate(dir);

                if (hasMultimeter(vector.intX(), vector.intY(), vector.intZ()))
                {
                    connections[dir.ordinal()] = getMultimeter(vector.intX(), vector.intY(), vector.intZ());
                }
            }
        }

        return connections;
    }

    @Override
    public IConnector<MultimeterNetwork> getInstance(ForgeDirection dir)
    {
        return this;
    }

    public universalelectricity.api.vector.Vector3 getPosition()
    {
        return new universalelectricity.api.vector.Vector3(x(), y(), z());
    }

    @Override
    @SideOnly(Side.CLIENT)
    public Cuboid6 getRenderBounds()
    {
        if (isPrimary)
            return Cuboid6.full.copy().expand(new Vector3(getNetwork().size.x, getNetwork().size.y, getNetwork().size.z));
        return Cuboid6.full;
    }

    @Override
    public String toString()
    {
        return "[PartMultimeter]" + x() + "x " + y() + "y " + z() + "z " + getSlotMask() + "s ";
    }

}

File: src/main/scala/edx/electrical/multimeter/PartMultimeter.scala
package edx.electrical.multimeter

import java.util.{ArrayList, HashSet, List, Set}

import codechicken.lib.data.{MCDataInput, MCDataOutput}
import codechicken.lib.vec
import codechicken.lib.vec.Cuboid6
import codechicken.multipart.{IRedstonePart, TMultiPart, TileMultipart}
import cpw.mods.fml.relauncher.{Side, SideOnly}
import edx.core.Electrodynamics
import edx.core.interfaces.TNodeMechanical
import edx.core.prefab.part.ChickenBonesWrapper._
import edx.core.prefab.part.PartFace
import edx.electrical.ElectricalContent
import io.netty.buffer.ByteBuf
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{ChatComponentText, MovingObjectPosition}
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids.{FluidTankInfo, IFluidHandler}
import resonantengine.api.tile.{INodeProvider, IRemovable}
import resonantengine.lib.mod.compat.energy.Compatibility
import resonantengine.lib.network.discriminator.PacketType
import resonantengine.lib.network.handle.IPacketReceiver
import resonantengine.lib.transform.vector.Vector3
import resonantengine.lib.utility.WrenchUtility

import scala.collection.convert.wrapAll._

/** Block that detects power.
  *
  * @author Calclavia
  */
class PartMultimeter extends PartFace with IRedstonePart with IPacketReceiver with IRemovable.ISneakWrenchable
{

  var playersUsing: Set[EntityPlayer] = new HashSet[EntityPlayer]
  /** Detection */
  var redstoneTriggerLimit: Double = .0
  var detectType: Byte = 0
  var graphType: Byte = 0
  var redstoneOn: Boolean = false
  var isPrimary: Boolean = false
  private var detectMode = DetectModes.NONE
  private var doDetect: Boolean = true
  private var grid: GridMultimeter = null

  override def preRemove()
  {
    if (!world.isRemote)
      getGrid.remove(this)
  }

  def updateDesc()
  {
    writeDesc(getWriteStream)
  }

  override def activate(player: EntityPlayer, part: MovingObjectPosition, item: ItemStack): Boolean =
  {
    if (WrenchUtility.isUsableWrench(player, player.inventory.getCurrentItem, x, y, z))
    {
      if (!this.world.isRemote)
      {
        doDetect = !doDetect
        player.addChatMessage(new ChatComponentText("Multimeter detection set to: " + doDetect))
        WrenchUtility.damageWrench(player, player.inventory.getCurrentItem, x, y, z)
      }
      return true
    }
    player.openGui(Electrodynamics, placementSide.ordinal, world, x, y, z)
    return true
  }

  override def update()
  {
    super.update()

    if (!world.isRemote)
    {
      if (doDetect) updateDetections
      val detectedValue = getGrid.graphs(detectType).getDouble()
      var outputRedstone = false

      detectMode match
      {
        case DetectModes.EQUAL =>
          outputRedstone = detectedValue == redstoneTriggerLimit
        case DetectModes.GREATER_THAN =>
          outputRedstone = detectedValue > redstoneTriggerLimit
        case DetectModes.GREATER_THAN_EQUAL =>
          outputRedstone = detectedValue >= redstoneTriggerLimit
        case DetectModes.LESS_THAN =>
          outputRedstone = detectedValue < redstoneTriggerLimit
        case DetectModes.LESS_THAN_EQUAL =>
          outputRedstone = detectedValue <= redstoneTriggerLimit
        case _ =>
      }

      getGrid.markUpdate
      if (ticks % 20 == 0)
      {
        if (outputRedstone != redstoneOn)
        {
          redstoneOn = outputRedstone
          tile.notifyPartChange(this)
        }
        updateGraph
      }
    }
    if (!world.isRemote)
    {
      for (player <- playersUsing)
      {
        updateGraph
      }
    }
  }

  def updateGraph()
  {
    sendPacket(2)
  }

  def updateDetections()
  {
    val receivingSide: ForgeDirection = getDirection.getOpposite
    val tileEntity: TileEntity = getDetectedTile

    if (tileEntity.isInstanceOf[INodeProvider])
    {
      val instance = ForgeDirection.values
        .map(dir => tileEntity.asInstanceOf[INodeProvider].getNode(classOf[TNodeMechanical], dir).asInstanceOf[TNodeMechanical])
        .headOption.orNull

      if (instance != null)
      {
        getGrid.torqueGraph.queue(instance.torque)
        getGrid.angularVelocityGraph.queue(instance.angularVelocity)
        getGrid.powerGraph.queue(instance.torque * instance.angularVelocity)
      }
    }
    if (tileEntity.isInstanceOf[IFluidHandler])
    {
      val fluidInfo: Array[FluidTankInfo] = tileEntity.asInstanceOf[IFluidHandler].getTankInfo(receivingSide)

      if (fluidInfo != null)
      {
        fluidInfo.filter(info => info != null && info.fluid != null).foreach(info => getGrid.fluidGraph.queue(info.fluid.amount))
      }
    }
    getGrid.energyGraph.queue(Compatibility.getEnergy(tileEntity, receivingSide))
  }

  def getDetectedTile: TileEntity =
  {
    val direction: ForgeDirection = getDirection
    return world.getTileEntity(x + direction.offsetX, y + direction.offsetY, z + direction.offsetZ)
  }

  override def write(packet: MCDataOutput, id: Int)
  {
    super.write(packet, id)

    id match
    {
      case 0 =>
      {
        packet.writeByte(placementSide.ordinal)
        packet.writeByte(facing)
        packet.writeByte(detectMode.id)
        packet.writeByte(detectType)
        packet.writeByte(graphType)
        packet.writeNBTTagCompound(getGrid.center.writeNBT(new NBTTagCompound))
        packet.writeNBTTagCompound(getGrid.size.writeNBT(new NBTTagCompound))
        packet.writeBoolean(getGrid.isEnabled)
      }
      case 2 =>
      {
        //Graph
        packet.writeByte(2)
        isPrimary = getGrid.isPrimary(this)
        packet.writeBoolean(isPrimary)
        if (isPrimary) packet.writeNBTTagCompound(getGrid.save)
      }
    }
  }

  override def read(packet: MCDataInput, packetID: Int)
  {
    packetID match
    {
      case 0 =>
      {
        placementSide = ForgeDirection.getOrientation(packet.readByte)
        facing = packet.readByte
        detectMode = DetectModes(packet.readByte).asInstanceOf[DetectModes.DetectMode]
        detectType = packet.readByte
        graphType = packet.readByte
        getGrid.center = new Vector3(packet.readNBTTagCompound)
        getGrid.size = new Vector3(packet.readNBTTagCompound)
        getGrid.isEnabled = packet.readBoolean
      }
      case 1 =>
      {
        redstoneTriggerLimit = packet.readLong
      }
      case 2 =>
      {
        isPrimary = packet.readBoolean
        if (isPrimary) getGrid.load(packet.readNBTTagCompound)
      }
    }
  }

  def getRemovedItems(entity: EntityPlayer): List[ItemStack] =
  {
    val list: List[ItemStack] = new ArrayList[ItemStack]
    list.add(new ItemStack(ElectricalContent.itemMultimeter))
    return list
  }

  def read(data: ByteBuf, player: EntityPlayer, `type`: PacketType)
  {
    detectMode = DetectModes(data.readByte).asInstanceOf[DetectModes.DetectMode]
    detectType = data.readByte
    graphType = data.readByte
    redstoneTriggerLimit = data.readDouble
  }

  def toggleGraphType
  {
    graphType = ((graphType + 1) % getGrid.graphs.size).asInstanceOf[Byte]
    updateServer
  }

  def toggleMode
  {
    detectMode = DetectModes((detectMode.id + 1) % DetectModes.values.size).asInstanceOf[DetectModes.DetectMode]
    updateServer
  }

  def updateServer
  {
  }

  def toggleDetectionValue
  {
    detectType = ((detectType + 1) % getGrid.graphs.size).asInstanceOf[Byte]
    updateServer
  }

  def getGrid: GridMultimeter =
  {
    if (grid == null)
    {
      grid = new GridMultimeter
      grid.add(this)
    }
    return grid
  }

  def setGrid(network: GridMultimeter)
  {
    grid = network
  }

  override def load(nbt: NBTTagCompound)
  {
    super.load(nbt)
    placementSide = ForgeDirection.getOrientation(nbt.getByte("side"))
    detectMode = DetectModes(nbt.getByte("detectMode")).asInstanceOf[DetectModes.DetectMode]
    detectType = nbt.getByte("detectionType")
    graphType = nbt.getByte("graphType")
    doDetect = nbt.getBoolean("doDetect")
    redstoneTriggerLimit = nbt.getDouble("triggerLimit")
  }

  override def save(nbt: NBTTagCompound)
  {
    super.save(nbt)
    nbt.setByte("side", placementSide.ordinal.asInstanceOf[Byte])
    nbt.setByte("detectMode", detectMode.id.asInstanceOf[Byte])
    nbt.setByte("detectionType", detectType)
    nbt.setByte("graphType", graphType)
    nbt.setBoolean("doDetect", doDetect)
    nbt.setDouble("triggerLimit", redstoneTriggerLimit)
  }

  def getMode = detectMode

  override def redstoneConductionMap: Int =
  {
    return 0x1F
  }

  override def solid(arg0: Int): Boolean =
  {
    return true
  }

  @SideOnly(Side.CLIENT)
  override def renderDynamic(pos: vec.Vector3, frame: Float, pass: Int)
  {
    if (pass == 0)
    {
      RenderMultimeter.render(this, pos.x, pos.y, pos.z)
    }
  }

  def canConnectRedstone(arg0: Int): Boolean =
  {
    return true
  }

  def strongPowerLevel(arg0: Int): Int =
  {
    return if (redstoneOn) 14 else 0
  }

  def weakPowerLevel(arg0: Int): Int =
  {
    return if (redstoneOn) 14 else 0
  }

  def canConnect(direction: ForgeDirection, obj: AnyRef): Boolean =
  {
    return obj.isInstanceOf[PartMultimeter]
  }

  def getConnections: Array[AnyRef] =
  {
    val connections: Array[AnyRef] = new Array[AnyRef](6)
    for (dir <- ForgeDirection.VALID_DIRECTIONS)
    {
      if (dir != getDirection && dir != getDirection.getOpposite)
      {
        val vector: Vector3 = getPosition.add(dir)
        if (hasMultimeter(vector.xi, vector.yi, vector.zi))
        {
          connections(dir.ordinal) = getMultimeter(vector.xi, vector.yi, vector.zi)
        }
      }
    }
    return connections
  }

  def getDirection: ForgeDirection =
  {
    return ForgeDirection.getOrientation(this.placementSide.ordinal)
  }

  def hasMultimeter(x: Int, y: Int, z: Int): Boolean =
  {
    return getMultimeter(x, y, z) != null
  }

  /** Gets the multimeter on the same plane. */
  def getMultimeter(x: Int, y: Int, z: Int): PartMultimeter =
  {
    val tileEntity: TileEntity = world.getTileEntity(x, y, z)

    if (tileEntity.isInstanceOf[TileMultipart])
    {
      val part: TMultiPart = (tileEntity.asInstanceOf[TileMultipart]).partMap(placementSide.ordinal)
      if (part.isInstanceOf[PartMultimeter])
      {
        return part.asInstanceOf[PartMultimeter]
      }
    }

    return null
  }

  def getPosition: Vector3 =
  {
    return new Vector3(x, y, z)
  }

  @SideOnly(Side.CLIENT)
  override def getRenderBounds: Cuboid6 =
  {
    if (isPrimary) return Cuboid6.full.copy.expand(new Vector3(getGrid.size.x, getGrid.size.y, getGrid.size.z))
    return Cuboid6.full
  }

  override def toString: String = "[PartMultimeter]" + x + "x " + y + "y " + z + "z " + getSlotMask + "s "

  protected def getItem: ItemStack =
  {
    return new ItemStack(ElectricalContent.itemMultimeter)
  }

  object DetectModes extends Enumeration
  {

    val NONE = DetectMode("none")
    val LESS_THAN = DetectMode("lessThan")
    val LESS_THAN_EQUAL = DetectMode("lessThanOrEqual")
    val EQUAL = DetectMode("equal")
    val GREATER_THAN_EQUAL = DetectMode("greaterThanOrEqual")
    val GREATER_THAN = DetectMode("greaterThan")

    case class DetectMode(display: String) extends super.Val(nextId)

  }

}